python/cuml/test/test_pickle.py

```diff
@@ -1,187 +1,328 @@
+
+    assert array_equal(D_before, D_after)
+    assert array_equal(I_before, I_after)
+
+
+@pytest.mark.parametrize('datatype', [np.float32, np.float64])
+@pytest.mark.parametrize('nrows', [unit_param(20)])
+@pytest.mark.parametrize('ncols', [unit_param(3)])
+@pytest.mark.parametrize('k', [unit_param(3)])
+def test_neighbors_pickle_nofit(tmpdir, datatype, nrows, ncols, k):
+
+    """
+    Note: This test digs down a bit far into the
+    internals of the implementation, but it's
+    important that regressions do not occur
+    from changes to the class.
+    """
+
+    model = cuml.neighbors.NearestNeighbors()
+
+    unpickled = pickle_save_load(tmpdir, model)
+
+    state = unpickled.__dict__
+
+    print(str(state))
+
+    assert state["n_indices"] == 0
+    assert "X_m" not in state
+    assert state["sizes"] is None
+    assert state["input"] is None
+
+    X_train, _, X_test = make_dataset(datatype, nrows, ncols)
+
+    model.fit(X_train)
+
+    unpickled = pickle_save_load(tmpdir, model)
+
+    state = unpickled.__dict__
+
+    assert state["n_indices"] == 1
+    assert "X_m" in state
+    assert state["sizes"] is not None
+    assert state["input"] is not None
+
+
+@pytest.mark.parametrize('datatype', [np.float32, np.float64])
+@pytest.mark.parametrize('nrows', [unit_param(20)])
+@pytest.mark.parametrize('ncols', [unit_param(3)])
+@pytest.mark.parametrize('k', [unit_param(3)])
+@pytest.mark.xfail(strict=True)
+def test_neighbors_mg_fails(tmpdir, datatype, nrows, ncols, k):
+
+    model = cuml.neighbors.NearestNeighbors()
+    model.n_indices = 2
+
+    pickle_save_load(tmpdir, model)
+
+
+@pytest.mark.parametrize('datatype', [np.float32, np.float64])
+@pytest.mark.parametrize('model', dbscan_model.values())
+@pytest.mark.parametrize('nrows', [unit_param(20)])
+@pytest.mark.parametrize('ncols', [unit_param(3)])
+def test_dbscan_pickle(tmpdir, datatype, model, nrows, ncols):
+    X_train, _, _ = make_dataset(datatype, nrows, ncols)
+
+    cu_before_pickle_predict = model.fit_predict(X_train).to_array()
+
+    cu_after_pickle_model = pickle_save_load(tmpdir, model)
+
+    cu_after_pickle_predict = cu_after_pickle_model.fit_predict(
+                              X_train
+                              ).to_array()
+
+    assert array_equal(cu_before_pickle_predict, cu_after_pickle_predict)
+
+
+@pytest.mark.parametrize('datatype', [np.float32, np.float64])
+@pytest.mark.parametrize('nrows', [unit_param(20)])
+@pytest.mark.parametrize('ncols', [unit_param(3)])
+def test_tsne_pickle(tmpdir, datatype, nrows, ncols):
+    iris = load_iris()
+    iris_selection = np.random.RandomState(42).choice(
+        [True, False], 150, replace=True, p=[0.75, 0.25])
+    X = iris.data[iris_selection]
+
+    model = cuml.manifold.TSNE()
+    
+    model = pickle_save_load(tmpdir, model)
+
+    Y = model.fit_transform(X)
+
+    cu_trust_after = trustworthiness(X, Y, 10)
+
+    assert cu_trust_after >= 0.9
```

Comment:
It would probably be better to verify that the trustworthiness before and after are the same, rather than verifying that it is >= .9.

Reply:
Currently there is only fit_transform, and in fit, I do store it in self.Y.
However the issue with TSNE is for each run, even though random_state is set, the output is different since TSNE is very sensitive to even small changes in bits, so checking if before == after will always fail.

But ye based on you suggestion, I am now checking the trustworthiness before and after pickling, and checking if both are good.

Reply:
I see you are storing the embeddings as ‘self.YY’ on the object after fit is called. 

In this case, I would verify only 2 things in this test:
1. Perform a fit before pickling and verify that self.YY after matches self.YY before
2. Do not perform a fit before picking and verify the parameter dict before matches after 

UMAP makes this a little easier, since it does have a transform function, so I only needed to test the trustworthiness. TSNE is a little more like dbscan in that the internal state should really be tested directly.

Reply:
Yep good ideas! Will do this pickling test :)